SwitchSpeak-iOS/SwitchSpeak-iOS/TouchGrid/TouchGrid.swift

```diff
@@ -84,53 +86,35 @@ class TouchGrid {
             removeSubTreeFromView(childnode)
         }
     }
+	
+
+	func beginScanning() {
+		scanningTimer = Timer.scheduledTimer(withTimeInterval: GlobalSettings.getUserSettings().scanSpeed.rawValue, repeats: true) { [weak self] _ in
+			self?.selectSubTree()
+		}
+	}
+	
+	func stopScanning() {
+		scanningTimer?.invalidate()
+	}
+	
     
     /*
-     * Highlights the childNumber'th child node of the curNode.
+     * Highlights the next child node of the curNode.
      */
     func selectSubTree() {
-        curNode.childNodes[childNumber].highlightSubTree()
-        let previousCurNode: Node = curNode
-		var timeDelay = 0.0
-		if (!curNode.childNodes[childNumber].dummy) {
-			timeDelay = GlobalSettings.getUserSettings().scanSpeed.rawValue
+
+		if curNode.childNodes.count > 0 {
+			self.previousCurNode.unHighlightSubTree()
+			childNumber = nextChildNumber
+			self.curNode.childNodes[childNumber].highlightSubTree()
+			self.previousCurNode = curNode
+			repeat {		//	repeat until we find a non-dummy node
+				nextChildNumber = (nextChildNumber + 1) % self.curNode.childNodes.count
```

Comment:
I suggest you actually remove the modulo here. If there are no dummy nodes, with the modulo, this will just go into an infinite loop. If a node only has dummy nodes as children, this is an error and should cause the program to crash.

Reply:
I added modulo so that we can begin scanning from the first child node of `curNode` once all its children are scanned. If there are no dummy nodes, wouldn't the `while` check fail after first iteration making the repeat-while loop end? 

You are right, I'm assuming here that at least one non-dummy child node is present. This assumption holds in our code since `curNode` is always a non-dummy node and hence it has at least one non-dummy child node. Even then to avoid an infinite loop, I'll put an extra condition in the `if` statement at line number 108 to count the number of non-dummy child nodes in `curNode`.